tools/dds/dds-sniffer/rs-dds-sniffer.cpp

```diff
@@ -181,88 +208,175 @@ void dds_sniffer::on_writer_removed( librealsense::dds::dds_guid guid, const cha
+    if( _print_discoveries )
     {
         print_reader_discovered( guid, topic_name, true );
     }
     save_topic_reader( guid, topic_name );
 }
 
-void dds_sniffer::on_reader_removed( librealsense::dds::dds_guid guid, const char* topic_name )
+void dds_sniffer::on_reader_removed( librealsense::dds::dds_guid guid, const char * topic_name )
 {
-    if (_print_discoveries)
+    if( _print_discoveries )
     {
         print_reader_discovered( guid, topic_name, false );
     }
 
     remove_topic_reader( guid, topic_name );
 }
 
-void dds_sniffer::on_participant_added( librealsense::dds::dds_guid guid, const char* participant_name )
+void dds_sniffer::on_participant_added( librealsense::dds::dds_guid guid, const char * participant_name )
 {
-    if (_print_discoveries)
+    if( _print_discoveries )
     {
         print_participant_discovered( guid, participant_name, true );
     }
 
-    std::lock_guard<std::mutex> lock( _dds_entities_lock );
+    std::lock_guard< std::mutex > lock( _dds_entities_lock );
 
     _discovered_participants[guid] = participant_name;
 }
-void dds_sniffer::on_participant_removed( librealsense::dds::dds_guid guid, const char* participant_name )
+
+void dds_sniffer::on_participant_removed( librealsense::dds::dds_guid guid, const char * participant_name )
 {
     if( _print_discoveries )
     {
         print_participant_discovered( guid, participant_name, false );
     }
 
-    std::lock_guard<std::mutex> lock( _dds_entities_lock );
+    std::lock_guard< std::mutex > lock( _dds_entities_lock );
 
     _discovered_participants.erase( guid );
 }
 
-void dds_sniffer::save_topic_writer( librealsense::dds::dds_guid guid, const char* topic_name )
+void dds_sniffer::on_type_discovery( char const * topic_name, DynamicType_ptr dyn_type )
+{
+    // Register type with participant
+    TypeSupport type_support( new DynamicPubSubType( dyn_type ) );
+    type_support.register_type( _participant.get() );
+    std::cout << "Discovered topic " << topic_name << " of type: " << type_support->getName() << std::endl;
+
+    if (_print_topic_samples)
+    {
+        // Create subscriber, topic and reader to receive instances of this topic
+        if (_discovered_types_subscriber == nullptr)
+        {
+            _discovered_types_subscriber = _participant.get()->create_subscriber( SUBSCRIBER_QOS_DEFAULT, nullptr );
+            if (_discovered_types_subscriber == nullptr)
+            {
+                std::cout << "Cannot create subscriber for discovered type '" << topic_name << std::endl;
+                return;
+            }
+        }
+
+        Topic* topic = _participant.get()->create_topic( topic_name, type_support->getName(), TOPIC_QOS_DEFAULT );
+        if (topic == nullptr)
+        {
+            std::cout << "Cannot create topic for discovered type '" << topic_name << std::endl;
+            return;
+        }
+
+        StatusMask sub_mask = StatusMask::subscription_matched() << StatusMask::data_available();
```

Comment:
Why not use `|` instead of `<<`?

Reply:
They override the `<<` operator, so maybe it's better to do this there way so if they change something in the future we will not break it?

```
    inline StatusMask& operator >>(
            const StatusMask& mask)
    {
        *this &= ~mask;
        return *this;
    }
```
    inline StatusMask& operator <<(
            const StatusMask& mask)
    {
        *this |= mask;
        return *this;
    }
```

```

Reply:
I don't even know how to read what they did.
I think the meaning (in both `<<` and `>>`) is "add flag" and "remove flag" but in that case (and since they inherit from `bitset`) I think `|` and `& ~` is clearer. But up to you :).

Reply:
IMHO, using `+` and `-` would have been MUCH clearer. Not sure why they used `<<` and `>>`...

Reply:
I agree that `<<` and `>>` are not very clear, but I will keep it their way so internal changes to `StatusMask` won't break our code.